rfcs/0004-fulltext-search.md

```diff
@@ -0,0 +1,236 @@
+    can be easily defined by a subgraph developer ultimately providing 
+    the tools necesssary for front end developers to create effective 
+    user interfaces without the need to middleware.
+     
+## Urgency
+
+A delay in adding the fulltext search feature  will not create issues with 
+current deployments. However, it will represent a realization of part of 
+the long term vision for the query network. Also several high profile users 
+have communicated that it may be a conversion blocker, so implementation 
+should be prioritized. 
+
+## Terminology
+
+- `lexeme`: a basic lexical unit of a language, consisting of one word or 
+  several words, considered as an abstract unit, and applied to a family 
+  of words related by form or meaning.
+- `morphology (linguistics)`: the study of words, how they are formed, 
+  and their relationship to other words in the same language. 
+- `fulltext search index`: the result of lexical and morphological 
+  analysis (stemming) of a set of text documents.  It provides frequency 
+  and location for the language-specific stems found in the text documents 
+  being indexed. 
+
+## Detailed Design
+
+### Subgraph Definition
+
+Part of the power of fulltext search indexes are their flexibility, so 
+it is important to expose a simple interface to facilitate useful 
+applications of the index and ulimately aim to reduce the need to create 
+new subgraphs for the express purpose of updating fulltext search fields 
+for end-use cases. 
+
+For each fulltext search index a subgraph developer must be able to specify:
+    1. a language, 
+    2. a set of text document fields to include,
+    3. relative weighting for each field in the index.
+
+The proposed process of adding a fulltext search index to an Entity 
+involves adding a special field to that Entity in the GraphQL schema of 
+the subgraph's definition. The subgraph developer will create a new virtual 
+field of type String that is marked with the special `@index` directive. 
+The directive will have two required parameters for specifying the index
+`type` and `params`. Params fields are specific to the index type; in the 
+case of fulltext search there will be two params to specify: `language` 
+and `fields`. Verification of the index definition will ensure that all 
+fields referenced are valid String type fields. To be clear, the proposed 
+interface will add the fulltext search index definitions to the subgraph 
+deployment hash, so adding or updating a fulltext search index will require 
+an update to the subgraph.  With subgraph composition it will be possible 
+to easily create new subgraphs that add specific fulltext search indexes 
+to an existing subgraph. 
+
+Example fulltext search field definition:
+```graphql
+type ExampleItem @entity {
+    id: ID!,
+    name: String!,
+    description: String!,
+    specs: String!,
+    address: Bytes!,
+    marked: Boolean!,
+    itemSearch: String! @index(
```

Comment:
If fulltext search is expected to work across entities, then this feels wrong. One idea could be to define indexes globally on the `_Schema_` type that we're introducing for e.g. schema composition:
```graphql
type _Schema_
  @index(
    index: {
      name: "search",
      language: "english",
      include: [
        { entity: "ExampleItem", fields: [ {name: "name", weight: 5 }, ... ] }
      ]
    }
  )
```

Note: In GraphQL, there is no such thing as an any type, which is why it is better not to decouple `type` and `params` and instead define possible indexes with something like
```graphql
type IndexDefinition = FulltextIndexDefinition | OtherIndexDefinition

type FulltextIndexEntityField {
  name: String!
  weight: Int!
}

type FulltextIndexEntity {
  entity: String!
  fields: [FulltextIndexEntityField!]!
}

type FulltextIndexDefinition {
  name: String!
  language: String!
  entities: [FulltextIndexEntity!]!,
}
```